exact_multiple_string_matching/commentz_walter.py

```diff
@@ -0,0 +1,184 @@
+from dataclasses import dataclass
+from collections import deque
+
+@dataclass
+class Node():
+  def __init__(self, char, depth, parent):
+    self.char = char
+    self.depth = depth
+    self.word = None
+    self.parent = parent
+    self.ac_suffix = None
+    self.ac_output = None
+    self.children = {}
+
+@dataclass
+class CommentzWalterNode(Node):
+  def __init__(self, char, depth, parent):
+    Node.__init__(self, char, depth, parent)
+    self.min_diff_s1 = -1
+    self.min_diff_s2 = -1
+    self.cw_suffix = None
+    self.cw_output = None
+    self.shift1 = None
+    self.shift2 = None
+
+# pylint: disable=no-self-use
+class Trie():
+  def create_node(self, char, depth, parent):
+    return Node(char, depth, parent)
+
+  def __init__(self):
+    self.size = 0
+    self.root = self.create_node(None, 0, None)
+
+  def add_word(self, word):
+    curr_node = self.root
+    curr_depth = 1
+    for char in word:
+      next_node = curr_node.children.get(char)
+      if not next_node:
+        next_node = self.create_node(char, curr_depth, curr_node)
+        curr_node.children[char] = next_node
+
+      curr_node = next_node
+      curr_depth += 1
+
+    if curr_node.word is not None:
+      return
+
+    curr_node.word = word
+    self.size += 1
+
+class CommentzWalterAutomaton(Trie):
+  def create_node(self, char, depth, parent):
+    return CommentzWalterNode(char, depth, parent)
+
+  def __init__(self):
+    Trie.__init__(self)
+    self.min_depth = None
+    self.char_arr = {}
+
+  def add_word(self, word):
+    word = word[::-1]
+    super().add_word(word)
+    pos = 1
+
+    for char in word:
+      min_char_depth = self.char_arr.get(char)
+      if (min_char_depth is None) or (min_char_depth > pos):
+        self.char_arr[char] = pos
+      pos += 1
+
+    if self.min_depth is None or len(word) < self.min_depth:
+      self.min_depth = len(word)
+
+  def init_shift_values(self):
+    bfs_queue = deque()
+    self.root.shift1 = 1
+    self.root.shift2 = self.min_depth
+
+    for key in self.root.children:
+      bfs_queue.append(self.root.children[key])
+
+    while bfs_queue:
+      curr_node = bfs_queue.popleft()
+
+      if curr_node.cw_suffix is None:
+        curr_node.shift1 = self.min_depth
+      else:
+        curr_node.shift1 = curr_node.min_diff_s1
+
+      if curr_node.cw_output is None:
+        curr_node.shift2 = curr_node.parent.shift2
+      else:
+        curr_node.shift2 = curr_node.min_diff_s2
```

Comment:
curr_node.shift2 = (curr_node.parent.shift2 if curr_node.cw_output is None else curr_node.min_diff_s2)
Above similarly.